File: scripts/pitch-fetch-descript-transcripts.mjs
#!/usr/bin/env node
// Fetch actual transcript text from Descript share pages for the 5 still-deferred voices.
// Descript pages embed a signed URL to transcript.json; we follow it and extract paragraphs.

import { writeFileSync } from 'fs';

const targets = [
  { name: 'Nigel Alice',           url: 'https://share.descript.com/view/81eXhBcx1sq' },
  { name: 'Jackquann',             url: 'https://share.descript.com/view/iMkgYckmB6r' },
  { name: 'Laquisha',              url: 'https://share.descript.com/view/upRsPoO2uG8' },
  { name: 'Jay',                   url: 'https://share.descript.com/view/CBFA59vK0Cv' },
  { name: 'Rashad Gavin Isaacson', url: 'https://share.descript.com/view/8ywPFEnrUp4' },
];

const UA = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36';

async function fetchShare(shareUrl) {
  const r = await fetch(shareUrl, { headers: { 'User-Agent': UA } });
  const html = await r.text();
  // Find the transcript URL embedded in page state (HTML-escaped)
  const m = html.match(/"transcript"\s*:\s*\{\s*"url"\s*:\s*"([^"]+)"/);
  if (!m) return null;
  const transcriptUrl = m[1].replace(/&amp;/g, '&');
  return transcriptUrl;
}

async function fetchTranscriptJson(url) {
  const r = await fetch(url, { headers: { 'User-Agent': UA } });
  if (!r.ok) throw new Error('transcript.json ' + r.status);
  return r.json();
}

function extractText(tx) {
  // Descript transcript.json shape varies. Common shapes:
  //   { paragraphs: [{ words: [{ text }] }] }
  //   { blocks: [{ children: [{ text }] }] }
  //   { text: "..." }
  if (tx.text && typeof tx.text === 'string') return tx.text;

  const paragraphs = [];

  // Shape 1: paragraphs[].words[]
  if (Array.isArray(tx.paragraphs)) {
    for (const p of tx.paragraphs) {
      const speaker = p.speaker_name || p.speaker || '';
      let line = '';
      if (Array.isArray(p.words)) line = p.words.map((w) => w.text || w.word || '').join('');
      else if (typeof p.text === 'string') line = p.text;
      if (line.trim()) paragraphs.push((speaker ? speaker + ': ' : '') + line.trim());
    }
    if (paragraphs.length) return paragraphs.join('\n\n');
  }

  // Shape 2: blocks[].children[]
  if (Array.isArray(tx.blocks)) {
    for (const b of tx.blocks) {
      if (Array.isArray(b.children)) {
        const line = b.children.map((c) => c.text || '').join('');
        if (line.trim()) paragraphs.push(line.trim());
      } else if (typeof b.text === 'string' && b.text.trim()) {
        paragraphs.push(b.text.trim());
      }
    }
    if (paragraphs.length) return paragraphs.join('\n\n');
  }

  // Fallback: dump keys so we can inspect
  return '(unrecognised transcript shape · keys=' + Object.keys(tx).join(',') + ')';
}

const out = [];
out.push('='.repeat(90));
out.push('DEFERRED CAST — Descript transcript text (fetched from share URLs)');
out.push('Generated ' + new Date().toISOString());
out.push('='.repeat(90));

for (const t of targets) {
  out.push('');
  out.push('#'.repeat(90));
  out.push('# ' + t.name);
  out.push('#'.repeat(90));
  out.push('source: ' + t.url);
  out.push('');

  try {
    const transcriptUrl = await fetchShare(t.url);
    if (!transcriptUrl) { out.push('(no transcript url in share page)'); continue; }
    const j = await fetchTranscriptJson(transcriptUrl);
    const text = extractText(j);
    out.push(text);
  } catch (e) {
    out.push('ERROR: ' + e.message);
  }
}

const path = 'data/pitch/descript-transcripts.txt';
writeFileSync(path, out.join('\n'));
console.log('Wrote', path);

File: scripts/pitch-fetch-descript-transcripts.py
#!/usr/bin/env python3
"""
Fetch raw Descript transcripts for the 5 still-deferred pitch voices.
Descript segments are word-level; we group by speaker into paragraphs.
"""

import re
import ssl
import json
import urllib.request
from pathlib import Path

UA = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36'
CTX = ssl.create_default_context()

TARGETS = [
    ('Nigel Alice',            'https://share.descript.com/view/81eXhBcx1sq'),
    ('Jackquann',              'https://share.descript.com/view/iMkgYckmB6r'),
    ('Laquisha',               'https://share.descript.com/view/upRsPoO2uG8'),
    ('Jay',                    'https://share.descript.com/view/CBFA59vK0Cv'),
    ('Rashad Gavin Isaacson',  'https://share.descript.com/view/8ywPFEnrUp4'),
]


def fetch(url):
    req = urllib.request.Request(url, headers={'User-Agent': UA})
    return urllib.request.urlopen(req, context=CTX).read().decode('utf-8', errors='ignore')


def find_transcript_url(share_html):
    m = re.search(r'"transcript"\s*:\s*\{\s*"url"\s*:\s*"([^"]+)"', share_html)
    return m.group(1).replace('&amp;', '&') if m else None


def assemble(transcript_json):
    """Group word-level segments into speaker-turn paragraphs."""
    segments = transcript_json.get('segments') or []
    if not segments:
        return '(no segments in transcript)'

    paragraphs = []
    current_speaker = None
    current_words = []

    def flush():
        if current_words:
            text = ' '.join(current_words).replace(' ,', ',').replace(' .', '.').replace(' ?', '?').replace(' !', '!')
            paragraphs.append(f'{current_speaker}: {text}')

    for seg in segments:
        sp = seg.get('speaker', 'Speaker')
        body = seg.get('body', '')
        if sp != current_speaker:
            flush()
            current_speaker = sp
            current_words = [body]
        else:
            current_words.append(body)
    flush()

    return '\n\n'.join(paragraphs)


out = []
out.append('=' * 90)
out.append('DEFERRED CAST — Descript transcripts (raw word-level reassembled)')
out.append('=' * 90)

for name, url in TARGETS:
    out.append('')
    out.append('#' * 90)
    out.append(f'# {name}')
    out.append('#' * 90)
    out.append(f'source: {url}')
    out.append('')

    try:
        html = fetch(url)
        tx_url = find_transcript_url(html)
        if not tx_url:
            out.append('(no transcript url in share page)')
            continue
        tx_json = json.loads(fetch(tx_url))
        out.append(assemble(tx_json))
    except Exception as e:
        out.append(f'ERROR: {e}')

Path('data/pitch/descript-transcripts.txt').write_text('\n'.join(out))
print('Wrote data/pitch/descript-transcripts.txt')
